File: libaums/src/com/github/mjdev/libaums/UsbMassStorageDevice.java
package com.github.mjdev.libaums;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import android.annotation.TargetApi;
import android.content.Context;
import android.hardware.usb.UsbConstants;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbEndpoint;
import android.hardware.usb.UsbInterface;
import android.hardware.usb.UsbManager;
import android.os.Build;
import android.util.Log;

import com.github.mjdev.libaums.driver.BlockDeviceDriver;
import com.github.mjdev.libaums.driver.BlockDeviceDriverFactory;
import com.github.mjdev.libaums.partition.Partition;
import com.github.mjdev.libaums.partition.PartitionTable;
import com.github.mjdev.libaums.partition.PartitionTableEntry;
import com.github.mjdev.libaums.partition.PartitionTableFactory;

public class UsbMassStorageDevice {
	
	@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
	private class JellyBeanMr2Communication implements UsbCommunication {
		@Override
		public int bulkOutTransfer(byte[] buffer, int length) {
			return deviceConnection.bulkTransfer(outEndpoint, buffer, length, TRANSFER_TIMEOUT);
		}
		
		@Override
		public int bulkOutTransfer(byte[] buffer, int offset, int length) {
			return deviceConnection.bulkTransfer(outEndpoint, buffer, offset, length, TRANSFER_TIMEOUT);
		}

		@Override
		public int bulkInTransfer(byte[] buffer, int length) {
			return deviceConnection.bulkTransfer(inEndpoint, buffer, length, TRANSFER_TIMEOUT);
		}

		@Override
		public int bulkInTransfer(byte[] buffer, int offset, int length) {
			return deviceConnection.bulkTransfer(inEndpoint, buffer, offset, length, TRANSFER_TIMEOUT);
		}
	}
	
	/**
	 * On Android API level lower 18 (Jelly Bean MR2) we cannot specify a start offset in the source/destination
	 * array. Because of that we have to use this workaround, where we have to copy the data every time offset is non zero.
	 * 
	 */
	private class HoneyCombMr1Communication implements UsbCommunication {
		@Override
		public int bulkOutTransfer(byte[] buffer, int length) {
			return deviceConnection.bulkTransfer(outEndpoint, buffer, length, TRANSFER_TIMEOUT);
		}
		
		@Override
		public int bulkOutTransfer(byte[] buffer, int offset, int length) {
			if(offset == 0)
				return deviceConnection.bulkTransfer(outEndpoint, buffer, length, TRANSFER_TIMEOUT);
			
			byte[] tmpBuffer = new byte[length];
			int result = deviceConnection.bulkTransfer(outEndpoint, tmpBuffer, length, TRANSFER_TIMEOUT);
			System.arraycopy(tmpBuffer, 0, buffer, offset, length);
			return result;
		}

		@Override
		public int bulkInTransfer(byte[] buffer, int length) {
			return deviceConnection.bulkTransfer(inEndpoint, buffer, length, TRANSFER_TIMEOUT);
		}

		@Override
		public int bulkInTransfer(byte[] buffer, int offset, int length) {
			if(offset == 0)
				return deviceConnection.bulkTransfer(inEndpoint, buffer, length, TRANSFER_TIMEOUT);
			
			byte[] tmpBuffer = new byte[length];
			int result = deviceConnection.bulkTransfer(inEndpoint, tmpBuffer, length, TRANSFER_TIMEOUT);
			System.arraycopy(tmpBuffer, 0, buffer, offset, length);
			return result;
		}
	}
	
	private static final String TAG = UsbMassStorageDevice.class.getSimpleName();
	
	// subclass 6 means that the usb mass storage device implements the
	// SCSI transparent command set
	private static final int INTERFACE_SUBCLASS = 6;
	// protocol 80 means the communication happens only via bulk transfers
	private static final int INTERFACE_PROTOCOL = 80;
	
	private static int TRANSFER_TIMEOUT = 21000;
	
	private UsbManager usbManager;
	private UsbDeviceConnection deviceConnection;
	private UsbDevice usbDevice;
	private UsbInterface usbInterface;
	private UsbEndpoint inEndpoint;
	private UsbEndpoint outEndpoint;
	
	private BlockDeviceDriver blockDevice;
	private PartitionTable partitionTable;
	private List<Partition> partitions = new ArrayList<Partition>();
	
	public UsbMassStorageDevice(UsbManager usbManager, UsbDevice usbDevice,
			UsbInterface usbInterface, UsbEndpoint inEndpoint,
			UsbEndpoint outEndpoint) {
		this.usbManager = usbManager;
		this.usbDevice = usbDevice;
		this.usbInterface = usbInterface;
		this.inEndpoint = inEndpoint;
		this.outEndpoint = outEndpoint;
	}

	public static UsbMassStorageDevice[] getMassStorageDevices(Context context) {
		UsbManager usbManager = (UsbManager) context.getSystemService(Context.USB_SERVICE);
		ArrayList<UsbMassStorageDevice> result = new ArrayList<UsbMassStorageDevice>();
		
		for(UsbDevice device : usbManager.getDeviceList().values()) {
			Log.i(TAG, "found usb device: " + device);
			
			int interfaceCount = device.getInterfaceCount();
			for(int i = 0; i < interfaceCount; i++) {
				UsbInterface usbInterface = device.getInterface(i);
				Log.i(TAG, "found usb interface: " + usbInterface);
				
				// we currently only support SCSI transparent command set with bulk transfers only!
				if(usbInterface.getInterfaceClass() != UsbConstants.USB_CLASS_MASS_STORAGE ||
						usbInterface.getInterfaceSubclass() != INTERFACE_SUBCLASS ||
						usbInterface.getInterfaceProtocol() != INTERFACE_PROTOCOL) {
					Log.i(TAG, "device interface not suitable!");
					continue;
				}
				
				int endpointCount = usbInterface.getEndpointCount();
				if(endpointCount != 2) {
					Log.w(TAG, "inteface endpoint count != 2");
				}
				
				UsbEndpoint outEndpoint = null;
				UsbEndpoint inEndpoint = null;
				for(int j = 0; j < endpointCount; j++) {
					UsbEndpoint endpoint = usbInterface.getEndpoint(j);
					Log.i(TAG, "found usb endpoint: " + endpoint);
					if(endpoint.getDirection() == UsbConstants.USB_DIR_OUT) {
						outEndpoint = endpoint;
					} else {
						inEndpoint = endpoint;
					}
				}
				
				if(outEndpoint == null || inEndpoint == null) {
					Log.e(TAG, "Not all needed endpoints found!");
					continue;
				}
				
				result.add(new UsbMassStorageDevice(usbManager, device, usbInterface, inEndpoint, outEndpoint));
				
			}
		}
		
		return result.toArray(new UsbMassStorageDevice[0]);
	}
	
	public void init() throws IOException {
		if(usbManager.hasPermission(usbDevice))
			setupDevice();
		else
			throw new IllegalStateException("Missing permission to access usb device: " + usbDevice);
			
	}
	
	private void setupDevice() throws IOException {
		Log.d(TAG, "setup device");
		deviceConnection = usbManager.openDevice(usbDevice);
		if(deviceConnection == null) {
			Log.e(TAG, "deviceConnetion is null!");
			return;
		}
		
		boolean claim = deviceConnection.claimInterface(usbInterface, true);
		if(!claim) {
			Log.e(TAG, "could not claim interface!");
			return;
		}
		
		UsbCommunication communication;
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
			communication = new JellyBeanMr2Communication();
		} else {
			Log.i(TAG, "using workaround usb communication");
			communication = new HoneyCombMr1Communication();
		}
		
		blockDevice = BlockDeviceDriverFactory.createBlockDevice(communication);
		blockDevice.init();
		partitionTable = PartitionTableFactory.createPartitionTable(blockDevice);
		initPartitions();
	}
	
	private void initPartitions() throws IOException {
		Collection<PartitionTableEntry> partitionEntrys = partitionTable.getPartitionTableEntries();
		
		for(PartitionTableEntry entry : partitionEntrys) {
			Partition partition = Partition.createPartition(entry, blockDevice);
			if(partition != null) {
				partitions.add(partition);
			}
		}
	}
	
	public void close() {
		Log.d(TAG, "close device");
		boolean release = deviceConnection.releaseInterface(usbInterface);
		if(!release) {
			Log.e(TAG, "could not release interface!");
		}
		deviceConnection.close();
	}
	
	public List<Partition> getPartitions() {
		return partitions;
	}

	public UsbDevice getUsbDevice() {
		return usbDevice;
	}
	
	public BlockDeviceDriver getScsiBlockDevice() {
		return blockDevice;
	}
}

File: libaums/src/main/java/com/github/mjdev/libaums/UsbMassStorageDevice.kt
/*
 * (C) Copyright 2014-2019 magnusja <github@mgns.tech>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 */

package com.github.mjdev.libaums

import android.content.Context
import android.hardware.usb.*
import android.util.Log
import com.github.mjdev.libaums.driver.BlockDeviceDriver
import com.github.mjdev.libaums.driver.BlockDeviceDriverFactory
import com.github.mjdev.libaums.driver.scsi.commands.sense.MediaNotInserted
import com.github.mjdev.libaums.partition.Partition
import com.github.mjdev.libaums.partition.PartitionTable
import com.github.mjdev.libaums.partition.PartitionTableFactory
import com.github.mjdev.libaums.usb.UsbCommunication
import com.github.mjdev.libaums.usb.UsbCommunicationFactory
import java.io.IOException

/**
 * Class representing a connected USB mass storage device. You can enumerate
 * through all connected mass storage devices via
 * [.getMassStorageDevices]. This method only returns supported
 * devices or if no device is connected an empty array.
 *
 *
 * After choosing a device you have to get the permission for the underlying
 * [android.hardware.usb.UsbDevice]. The underlying
 * [android.hardware.usb.UsbDevice] can be accessed via
 * [.getUsbDevice].
 *
 *
 * After that you need to call [.setupDevice]. This will initialize the
 * mass storage device and read the partitions (
 * [com.github.mjdev.libaums.partition.Partition]).
 *
 *
 * The supported partitions can then be accessed via [.getPartitions]
 * and you can begin to read directories and files.
 *
 * @author mjahnen
 */
class UsbMassStorageDevice
/**
 * Construct a new [com.github.mjdev.libaums.UsbMassStorageDevice].
 * The given parameters have to actually be a mass storage device, this is
 * not checked in the constructor!
 *
 * @param usbManager
 * @param usbDevice
 * @param usbInterface
 * @param inEndpoint
 * @param outEndpoint
 */
private constructor(private val usbManager: UsbManager,
                    /**
                     * This returns the [android.hardware.usb.UsbDevice] which can be used
                     * to request permission for communication.
                     *
                     * @return Underlying [android.hardware.usb.UsbDevice] used for
                     * communication.
                     */
                    val usbDevice: UsbDevice,
                    private val usbInterface: UsbInterface, private val inEndpoint: UsbEndpoint, private val outEndpoint: UsbEndpoint) {

    lateinit var partitions: List<Partition>

    // TODO this is never used, should we only allow one init() call?
    private var inited = false
    
    private lateinit var usbCommunication: UsbCommunication

    /**
     * Initializes the mass storage device and determines different things like
     * for example the MBR or the file systems for the different partitions.
     *
     * @throws IOException
     * If reading from the physical device fails.
     * @throws IllegalStateException
     * If permission to communicate with the underlying
     * [UsbDevice] is missing.
     * @see .getUsbDevice
     */
    @Throws(IOException::class)
    fun init() {
        if (usbManager.hasPermission(usbDevice))
            setupDevice()
        else
            throw IllegalStateException("Missing permission to access usb device: $usbDevice")

        inited = true
    }

    /**
     * Sets the device up. Claims interface and initiates the device connection.
     * Chooses the right[UsbCommunication]
     * depending on the Android version (
     * [com.github.mjdev.libaums.usb.HoneyCombMr1Communication]
     * or (
     * [com.github.mjdev.libaums.usb.JellyBeanMr2Communication]
     * ).
     *
     *
     * Initializes the [.blockDevice] and reads the partitions.
     *
     * @throws IOException
     * If reading from the physical device fails.
     * @see .init
     */
    @Throws(IOException::class)
    private fun setupDevice() {
        usbCommunication = UsbCommunicationFactory
                .createUsbCommunication(usbManager, usbDevice, usbInterface, outEndpoint, inEndpoint)
        val maxLun = ByteArray(1)
        usbCommunication.controlTransfer(161, 254, 0, usbInterface.id, maxLun, 1)

        Log.i(TAG, "MAX LUN " + maxLun[0].toInt())

        this.partitions = (0..maxLun[0])
                .map { lun ->
                    BlockDeviceDriverFactory.createBlockDevice(usbCommunication, lun = lun.toByte())
                }
                .mapNotNull { blockDevice ->
                    try {
                        blockDevice.init()
                    } catch (e: MediaNotInserted) {
                        // This LUN does not have media inserted. Ignore it.
                        return@mapNotNull null
                    }

                    val partitionTable = PartitionTableFactory.createPartitionTable(blockDevice)

                    initPartitions(partitionTable, blockDevice)
                }
                .flatten()
    }

    /**
     * Fills [.partitions] with the information received by the
     * [.partitionTable].
     *
     * @throws IOException
     * If reading from the [.blockDevice] fails.
     */
    @Throws(IOException::class)
    private fun initPartitions(partitionTable: PartitionTable, blockDevice: BlockDeviceDriver) =
            partitionTable.partitionTableEntries.mapNotNull {
                Partition.createPartition(it, blockDevice)
            }

    /**
     * Releases the [android.hardware.usb.UsbInterface] and closes the
     * [android.hardware.usb.UsbDeviceConnection]. After calling this
     * method no further communication is possible. That means you can not read
     * or write from or to the partitions returned by [.getPartitions].
     */
    fun close() {
        usbCommunication.close()
        inited = false
    }

    companion object {

        private val TAG = UsbMassStorageDevice::class.java.simpleName

        /**
         * subclass 6 means that the usb mass storage device implements the SCSI
         * transparent command set
         */
        private const val INTERFACE_SUBCLASS = 6

        /**
         * protocol 80 means the communication happens only via bulk transfers
         */
        private const val INTERFACE_PROTOCOL = 80

        @JvmStatic
        fun UsbDevice.getMassStorageDevices(context: Context): List<UsbMassStorageDevice> {
            val usbManager = context.getSystemService(Context.USB_SERVICE) as UsbManager

            return (0 until this.interfaceCount)
                    .map { getInterface(it) }
                    .filter {
                        // we currently only support SCSI transparent command set with
                        // bulk transfers only!
                        it.interfaceClass == UsbConstants.USB_CLASS_MASS_STORAGE
                                && it.interfaceSubclass == INTERFACE_SUBCLASS
                                && it.interfaceProtocol == INTERFACE_PROTOCOL
                    }
                    .map { usbInterface ->
                        Log.i(TAG, "Found usb interface: $usbInterface")

                        // Every mass storage device has exactly two endpoints
                        // One IN and one OUT endpoint
                        val endpointCount = usbInterface.endpointCount
                        if (endpointCount != 2) {
                            Log.w(TAG, "Interface endpoint count != 2")
                        }

                        var outEndpoint: UsbEndpoint? = null
                        var inEndpoint: UsbEndpoint? = null

                        for (j in 0 until endpointCount) {
                            val endpoint = usbInterface.getEndpoint(j)
                            Log.i(TAG, "Found usb endpoint: $endpoint")
                            if (endpoint.type == UsbConstants.USB_ENDPOINT_XFER_BULK) {
                                if (endpoint.direction == UsbConstants.USB_DIR_OUT) {
                                    outEndpoint = endpoint
                                } else {
                                    inEndpoint = endpoint
                                }
                            }
                        }

                        if (outEndpoint == null || inEndpoint == null) {
                            Log.e(TAG, "Not all needed endpoints found. In: ${outEndpoint != null}, Out: ${outEndpoint != null}")
                            return@map null
                        }

                        return@map UsbMassStorageDevice(
                                usbManager, this, usbInterface, inEndpoint, outEndpoint
                        )
                    }
                    .filterNotNull()
        }

        /**
         * This method iterates through all connected USB devices and searches for
         * mass storage devices.
         *
         * @param context
         * Context to get the [UsbManager]
         * @return An array of suitable mass storage devices or an empty array if
         * none could be found.
         */
        @JvmStatic
        fun getMassStorageDevices(context: Context): Array<UsbMassStorageDevice> {
            val usbManager = context.getSystemService(Context.USB_SERVICE) as UsbManager

            return usbManager.deviceList
                    .map {
                        val device = it.value
                        Log.i(TAG, "found usb device: $it")
                        device.getMassStorageDevices(context)
                    }
                    .flatten()
                    .toTypedArray()
        }
    }
}
